#!/usr/bin/env Rscript
#
# Copyright (C) 2018-2019 University of Glasgow
#
# Author: Dario Beraldi <dario.beraldi@glasgow.ac.uk>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

suppressMessages(library(argparse))
suppressMessages(library(data.table))
suppressMessages(library(Rsamtools))

# -----------------------------------------------------------------------------

VERSION= sprintf('0.16.0')

docstring<- sprintf('DESCRIPTION \\n\\
Detect somatic copy number variants (CNVs) and estimate purity and ploidy in a\\n\\
tumour sample given a matched normal sample. The core of the analysis is based\\n\\
on the package FACETS (https://github.com/mskcc/facets).\\n\\
\\n\\
INPUT\\n\\
\\n\\
* A bam file for the tumour sample, a bam file for the matched normal sample, and a vcf\\n\\
  file of SNP positions. Bam and vcf files must be sorted and indexed\\n\\
\\n\\
Or:\\n\\
\\n\\
* A csv file of counts (pileup) generated by a previous run of cnv_facets.R\\n\\
  or by the accompanying program snp-pileup\\n\\
\\n\\
EXAMPLE\\n\\
\\n\\
    cnv_facets.R -n normal.bam -t tumour.bam -vcf snps.vcf.gz -o my_cnv\\n\\
\\n\\
See the online documentation for more details and usage.\\n\\
Version %s', VERSION)

epilog<- 'Options --snp-* are used only to generate the pileup file. They are ignored with option --pileup'
                        
parser<- ArgumentParser(description= docstring, formatter_class= 'argparse.RawTextHelpFormatter', epilog= epilog)

parser$add_argument('--out-snp-pileup', '-op', help= 'Output path for snp-pileup file', required= TRUE)
parser$add_argument('--out-nbhd-snp', '-on', help= 'Output path for snp-pileup file', required= TRUE)

parser$add_argument('--snp-tumour', '-t', help= 'BAM file for tumour sample', required= FALSE)
parser$add_argument('--snp-normal', '-n', help= 'BAM file for normal sample', required= FALSE)

parser$add_argument('--snp-vcf', '-vcf', help= 'VCF file of SNPs where pileup is to be computed', required= FALSE)

def<- 5
parser$add_argument('--snp-mapq', '-mq', help= sprintf(
'Sets the minimum threshold for mapping quality. Default %s', def), required= FALSE, default= def, type= 'integer')

def<- 10
parser$add_argument('--snp-baq', '-bq', help= sprintf(
'Sets the minimum threshold for base quality. Default %s', def), required= FALSE, default= def, type= 'integer')

parser$add_argument('--snp-count-orphans', '-A', help= 
'Do not discard anomalous read pairs', action= 'store_true')

def<- 1
parser$add_argument('--snp-nprocs', '-N', help= sprintf(
'Number of parallel processes to run to prepare the read pileup file.\\n\\
Each chromsome is assigned to a process. Default %s', def), required= FALSE, default= def, type= 'integer')

def_nbhd<- list('auto', 250)
parser$add_argument('--nbhd-snp', '-snp', help= sprintf(
'If an interval of size nbhd-snp contains more than one SNP,\\n\\
sample a random one.  This sampling reduces the SNP serial\\n\\
correlation. This value should be similar to the median insert\\n\\
size of the libraries. If "auto" and if using paired end BAM\\n\\
input, use the estimated insert size from the normal bam file.\\n\\
Otherwise use 250. Default %s', def_nbhd[[1]]), type= 'character', default= def_nbhd[[1]])

# NB: argparse v1.1.1+ required for -v option to work.
parser$add_argument("-v", "--version", action= 'version', version= VERSION)

# -----------------------------------------------------------------------------
avg_insert_size<- function(bam, default){
    idx<- data.table(idxstatsBam(bam))[mapped > 0][order(-mapped)]
    if(nrow(idx) > 10){
        idx<- idx[1:10,]
    }
    ins_size<- c()
    n_reads<- c()
    hdr_len<- length(scanBamHeader(bam)[[1]][['text']])
    for(chrom in idx$seqnames){
        cmd<- sprintf('samtools view -q 3 -f 3 -F 3840 -h %s %s | head -n %s | samtools stats | grep ^SN | cut -f 2-',
                      bam, chrom, 200000 + hdr_len)
        stats<- system(cmd, intern= TRUE)
        size<- grep('insert size average', stats, value= TRUE)
        size<- as.numeric(unlist(strsplit(size, '\t'))[2])
        n<- grep('reads properly paired', stats, value= TRUE)
        n<- as.numeric(unlist(strsplit(n, '\t'))[2])
        if(n == 0){
            next 
        } 
        ins_size<- c(ins_size, size)
        n_reads<- c(n_reads, n)
    }
    if(length(n_reads) == 0){
        return(as.numeric(default))
    }
    return(weighted.mean(ins_size, n_reads))
}

exec_snp_pileup<- function(chrom, snp_vcf, output, normal_bam, tumour_bam, mapq, baq, pseudo_snp, keep_orphans){
    # Execute snp-pileup on chromosome `chrom` 
    # Send tmp output to the same directory of the final output so we are sure
    # we can write there. 
    d<- dirname(output)
    chrom_vcf<- file.path(d, paste0(sub('\\.vcf\\.gz$|\\.vcf\\.bgz$', '', basename(snp_vcf)), '.', chrom, '.vcf'))
    chrom_nbam<- file.path(d, paste0(sub('\\.bam', '', basename(normal_bam)), '.', chrom, '.bam'))
    chrom_tbam<- file.path(d, paste0(sub('\\.bam', '', basename(tumour_bam)), '.', chrom, '.bam'))

    if(keep_orphans) {
        orphans<- '--count-orphans'
    } else {
        orphans<- ''
    }
       
    cmd<- c('#!/bin/bash', '\n',
            '\n',
            'set -eo pipefail', '\n',
            '\n',
            'mkfifo', chrom_vcf, '\n',
            'mkfifo', chrom_nbam, '\n',
            'mkfifo', chrom_tbam, '\n',
            '\n',
            'bcftools view --output-type u', snp_vcf, chrom, '>', chrom_vcf, '&\n',
            'pid_bcf=$!', '\n',
            '\n',
            'samtools view -u', normal_bam, chrom, '>', chrom_nbam, '&\n',
            'pid_nbam=$!', '\n',
            '\n',
            'samtools view -u', tumour_bam, chrom, '>', chrom_tbam, '&\n',
            'pid_tbam=$!', '\n',
            '\n',
            'snp-pileup', 
            '--gzip',
            '--pseudo-snps', pseudo_snp,
            '--min-map-quality', mapq,
            '--min-base-quality', baq,
            '--max-depth 10000000',
            '--min-read-counts', '0,0',
            orphans,
            chrom_vcf, output, chrom_nbam, chrom_tbam,
            '\n',
            'set +e', '\n',
            '\n',
            'wait $pid_bcf; exit_code=$?', '\n',
            'if [[ $exit_code != 0 && $exit_code != 141 ]]; then exit $exit_code; fi', '\n',
            '\n',
            'wait $pid_nbam; exit_code=$?', '\n',
            'if [[ $exit_code != 0 && $exit_code != 141 ]]; then exit $exit_code; fi', '\n',
            '\n',
            'wait $pid_tbam; exit_code=$?', '\n',
            'if [[ $exit_code != 0 && $exit_code != 141 ]]; then exit $exit_code; fi', '\n'
       )
    cmd<- paste(cmd, collapse= ' ')
    script<- file.path(d, sprintf('snp_pileup.%s.sh', chrom))
    write(cmd, script)
    status<- system2("/bin/bash", args= script)
    if(status != 0){
        stop(sprintf('\nError in computing snp pileup. Exit code %s from execution of %s\n' , status, script))
    }
    unlink(c(chrom_vcf, chrom_nbam, chrom_tbam))
    return(cmd)
}

exec_snp_pileup_parallel<- function(snp_vcf, output, normal_bam, tumour_bam, mapq, baq, pseudo_snp, nprocs, keep_orphans){
    
    dtm<- format(Sys.time(), "%y%m%d-%H%M%S")
    tmpdir<- tempfile(pattern = paste0(basename(sub('\\.cvs\\.gz$', '', output)), '_', dtm, '_'), tmpdir = dirname(output))
    dir.create(tmpdir)
    
    chroms<- headerTabix(snp_vcf)$seqnames

    cl<- makeCluster(nprocs, type= 'FORK')
    chrom_csv<- parLapply(cl, chroms, function(chrom){
        chrom_csv<- file.path(tmpdir, paste0(chrom, '.csv.gz'))
        cmd<- exec_snp_pileup(chrom= chrom, 
                              snp_vcf= snp_vcf,
                              output= chrom_csv, 
                              normal_bam= normal_bam, 
                              tumour_bam= tumour_bam, 
                              mapq= mapq, 
                              baq= baq, 
                              pseudo_snp= pseudo_snp,
                              keep_orphans= keep_orphans)
        return(chrom_csv)
    })
    stopCluster(cl)
    concat_csv(chrom_csv, output, tmpdir= tmpdir)
    unlink(tmpdir, recursive= TRUE)
}

concat_csv<- function(csv_list, xfile, tmpdir){
    isGzip<- ifelse(grepl('\\.gz$', xfile), TRUE, FALSE)
    if(isGzip == TRUE){
        conn<- file.path(tmpdir, sub('\\.gz', '', basename(xfile))) 
    } else {
        conn<- xfile
    }
    col.names<- TRUE
    for(csv in csv_list){
        options(datatable.fread.input.cmd.message= FALSE)
        fwrite(x= fread(sprintf('gzip -d -c %s', csv)), file= conn, sep= ',', col.names= col.names, row.name= FALSE, quote= FALSE, append= !isTRUE(col.names))
        options(datatable.fread.input.cmd.message= TRUE)
        col.names<- FALSE
    }
    if(isGzip == TRUE){
        system2(c('gzip', '-c', conn), stdout= xfile)
        unlink(conn)
    }
}

if(sys.nframe() == 0){
    # Script is being executed from the command line

    # ---------------- [Validate arguments] -------------

    xargs<- parser$parse_args()

    if(is.null(xargs$snp_tumour) || is.null(xargs$snp_normal)){
        write('Please provide both a tumour and a normal bam file', stderr()) 
        quit(status= 1)
    }
    if(! is.null(xargs$snp_tumour) && xargs$snp_tumour == xargs$snp_normal){
        write('Input bam files for tumour and normal cannot be the same', stderr())
        quit(status= 1)
    }

    est_insert_size<- NA
    if(xargs$nbhd_snp == "auto"){
        if(!is.null(xargs$pileup)){
            nbhd_snp<- def_nbhd[[2]]
        } else {
            nbhd_snp<- avg_insert_size(xargs$snp_normal, default= def_nbhd[[2]])
            est_insert_size<- round(nbhd_snp, 2)
        }
    } else if(is.na(as.integer(xargs$nbhd_snp))){
        write('Argument to --nbhd-snp must be "auto" or an integer', stderr())
        quit(status= 1)
    } else {
        nbhd_snp<- as.integer(xargs$nbhd_snp)  
    }
    nbhd_snp<- as.integer(nbhd_snp)
    write.table(data.frame(nbhd_snp=nbhd_snp), xargs$out_nbhd_snp, quote=F, sep="\t", row.names=F)

    # --------------- [Run pileup] ----------------

    exec_snp_pileup_parallel(snp_vcf=xargs$snp_vcf, 
                             output=xargs$out_snp_pileup, 
                             normal_bam=xargs$snp_normal, 
                             tumour_bam=xargs$snp_tumour, 
                             mapq=xargs$snp_mapq, 
                             baq=xargs$snp_baq, 
                             pseudo_snp=nbhd_snp,
                             nprocs=ceiling(xargs$snp_nprocs / 2),
                             keep_orphans=xargs$snp_count_orphans)
}
